python/src/scipp/neutron/diffraction/load.py

```diff
@@ -0,0 +1,82 @@
+    LoadDiffCal. This algorithm produces up to three workspaces, a
+    TableWorkspace containing conversion factors between TOF and d, a
+    GroupingWorkspace with detector groups and a MaskWorkspace for masking.
+    The information from the TableWorkspace and GroupingWorkspace is converted
+    to a Scipp dataset and returned, while the MaskWorkspace is ignored for
+    now. Only the keyword paramters Filename and InstrumentName are mandatory.
+
+    Example of use:
+
+      from scipp.neutron.diffraction import load
+      cal = loadCal(Filename='cal_file.cal', InstrumentName='WISH')
+
+    Note that this function requires mantid to be installed and available in
+    the same Python environment as scipp.
+
+    :param str Filename: The name of the cal file to be loaded.
+    :param str InstrumentName: Name of Mantid instrument to use.
+    :param str InstrumentFilename: If specified, over-write the instrument
+                                    definition in the final Dataset with the
+                                    geometry contained in the file.
+    :param str GroupFilename: If specified, over-writes the grouping from
+                               CallFileName.
+    :param float TofMin: Minimum time of flight (default 0)
+    :param float TofMax: Maximum time of flight
+    :raises: If the Mantid workspace type returned by the Mantid loader is not
+             either EventWorkspace or Workspace2D.
+    :return: A Dataset containing the calibration data and grouping.
+    :rtype: Dataset
+    """
+
+    try:
+        import mantid.simpleapi as mantid
+    except ImportError as e:
+        raise ImportError(
+            "Mantid Python API was not found, please install Mantid framework "
+            "as detailed in the installation instructions (https://scipp."
+            "readthedocs.io/en/latest/getting-started/installation.html)"
+        ) from e
+
+    output = mantid.LoadDiffCal(Filename=Filename,
+                                InstrumentName=InstrumentName,
+                                WorkspaceName="cal_output",
+                                InstrumentFilename=InstrumentFilename,
+                                GroupFilename="",
+                                TofMin=TofMin, TofMax=TofMax)
+
+    cal_ws = output.OutputCalWorkspace
+    cal_data = sc.compat.mantid.convert_TableWorkspace_to_dataset(cal_ws)
+
+    # Modify units of cal_data
+    cal_data["difc"].unit = sc.units.us/sc.units.angstrom
+    # Currently unsupported unit on python 3.6
+    # cal_data["difa"].unit = sc.units.us/sc.units.angstrom/sc.units.angstrom
+    cal_data["tzero"].unit = sc.units.us
+
+    # Mask data not used yet
+    # output.OutputMaskWorkspace)
+
+    group_ws = output.OutputGroupingWorkspace
+
+    group_list = []
+    for i in range(group_ws.getNumberHistograms()):
+        group_list.append(group_ws.readY(i))
+
+    group_var = sc.Variable([sc.Dim.Row], values=group_list)
+
+    cal_data["group"] = group_var
+
```

Comment:
Reading this it occurs to me that we kind of have a "leak" in the load algorithms: We are not deleting the Mantid workspaces, so I suppose they persist? Can you check? We probably have to remove them from the `AnalysisDataService` by hand?

Reply:
I am not quite sure how I would test this? Any pointers on how to approach this?

Reply:
Can you just try to access it and see if there are any workspaces in it? Something like `len(mantid.mtd)` (may need to do `import mantid.simpleapi as mantid` first, since you original import is in a function? I am not quite clear on how Python works in this respect).

Reply:
So there is definitely a leak as you suspected. I can access the workspaces from outside the load_calibration function.
There is a clear method in the mantid.mtd class that seems like a good way to end the load functions, but I am not sure about if it actually deletes the objects in memory, or just forgets about them. I only see a smal drop in memory usage when invoking the clear method.

Reply:
I think we also have to be careful to not delete everything: Someone might be using Mantid directly (in addition to using it via this function), so we should remove only the workspaces created by this algorithm.

Regarding memory usage not dropping: This may not mean much. Even if it is free'ed, the pages might not be handed back to the OS immediately.

Reply:
Good point. I remove the workspaces created by LoadDiffCal and nothing else. It frees 123MB RAM immediately with the WISH calibration file.